_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+1. The name must be unique, so that it prevent naming collision
+2. It needs to support reloading modules, so the name needs to be deterministic so that it can be replaced
+
+Let's look at [the method I came up with](https://github.com/rapid7/metasploit-framework/blob/8a2dc0a09fbe4ed9880b515d36df751042b09d57/lib/msf/core/module_manager.rb#L1128-L1185)
+
+![`Msf::ModuleManager.wrapper_module`](/images/msf-module-manager-wrapper-module.png)
+
+`name` when it comes in is `/` separated, such as `auxiliary/pro/social_engineering/web_phish`.  I know from Rails that `String#camelize` in `activesupport` can be used to convert a path to a `Module#name`.
+
+![`Msf::ModuleManager.wrapper_module` `name` argument](/images/msf-module-manager-wrapper-module-name.png)
+
+So, `relative_module_name`, is `'Auxiliary::Pro::SocialEngineering::WebPhish'`
+
+![`Msf::ModuleManager.wrapper_module` `relative_module_name` example](/images/msf-module-manager-wrapper-module-relative-module-name.png)
+
+Because this now gives us a hierarchy of namespace modules, we need to ensure that each parent module is created before any child modules
+
+![`Msf::ModuleManager.wrapper_module` `until` loop](/images/msf-module-manager-wrapper-module-until.png)
+
+First, why am I starting with `Object` as the parent?  Well, it's because constant resolution all start with `Object.get_const`, which I got from `ModuleConstMissing#const_missing` (https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L176)
+
+![`ModuleConstMissing#const_missing` using `Object` as constant resolution root](/images/module-const-missing-const-missing-object.png)
+
+`until` is to `while` as `unless` is to `if`, so this `until` loop is walking down the descendants until it gets to a leaf, which must be the wrapper_module that should be returned.
+
+There are some slight complications that were encountered that weren't in the original constraints: some modules have components of their name which are leading digits or non-alphanumeric characters, such as '-'. To ensure a unique mapping, I hex encoded those illegal characters.  Since hex encoding is reversible, I know it is a 1-to-1 mapping and will maintain the uniqueness of the original names.  Hex encoding could return a digit again, which isn't allowed in a Module name, so I have to prefix hex encoded characters with `X`
+
+![`Msf::ModuleManager.wrapper_module` hex-encoding](/images/msf-module-manager-wrapper-module-hex-encoding.png)
+
+Module reloading is handled by removing the leaf constant if it already exist when the loop gets to it (`module_names.empty?` is `true`).    `false` is passed to `const_defined?` to speed constant lookup and to prevent false positives from constants defined in ancestors.
```

Comment:
Module reloading is handled by removing the leaf constant if it already **~~exist~~** **exists** when the loop gets to it (`module_names.empty?` is `true`).    `false` is passed to `const_defined?` to speed constant lookup and to prevent false positives from constants defined in ancestors.